client/client-lib/src/api.rs

```diff
@@ -1,171 +1,575 @@
+    E: From<ApiError>,
+{
+    let params_raw = match serde_json::to_value(params)
+        .map_err(Into::into)
+        .map_err(ApiError::ParamsSerialization)?
+    {
+        Value::Array(v) => v,
+        o => vec![o],
+    };
+
+    Ok(f(params_raw).await?)
+}
+
+#[cfg_attr(target_family = "wasm", async_trait(? Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+pub trait IFederationApiFed: IFederationApi {
+    async fn request_fed_with_strategy<MemberRet: serde::de::DeserializeOwned, FedRet>(
         &self,
-        epoch: u64,
-        epoch_pk: PublicKey,
-    ) -> Result<SignedEpochOutcome>;
+        mut strategy: impl QueryStrategy<MemberRet, FedRet> + Send,
+        method: &str,
+        params: &[Value],
+    ) -> FedResult<FedRet> {
+        let mut futures = FuturesUnordered::new();
 
-    async fn fetch_last_epoch(&self) -> Result<u64>;
+        let peers = self.members();
 
-    // TODO: more generic module API extensibility
-    /// Fetch ln contract state
-    async fn fetch_contract(&self, contract: ContractId) -> Result<ContractAccount>;
+        for peer_id in peers {
+            futures.push(Box::pin(async {
+                PeerResponse {
+                    peer: *peer_id,
+                    result: self.request_raw(*peer_id, method, params).await,
+                }
+            }) as Pin<Box<dyn Future<Output = _> + Send>>);
+        }
+
+        let mut peer_errors = BTreeMap::new();
+
+        // Delegates the response handling to the `QueryStrategy` with an exponential back-off
+        // with every new set of requests
+        let mut delay_ms = 10;
+        let max_delay_ms = 1000;
+        loop {
+            match futures.next().await {
+                Some(PeerResponse { peer, result }) => {
+                    // brain fog, put me out of my type system misery and rewrite this chunk
+                    let result: Result<MemberRet> = match result {
+                        Ok(o) => match serde_json::from_value(o) {
+                            Ok(o) => Ok(o),
+                            Err(e) => Err(ApiError::ResponseDeserialization(e.into())),
+                        },
+                        Err(e) => {
+                            peer_errors.insert(peer, e);
+                            Err(e)
+                        }
+                    };
+                    match strategy.process(PeerResponse { peer, result }) {
+                        QueryStep::Retry(peers) => {
+                            for retry_peer in peers {
+                                peer_errors.remove(&retry_peer);
+                                futures.push(Box::pin(async {
+                                    // Note: we need to sleep inside the retrying future,
+                                    // so that `futures` is being polled continously
+                                    sleep(Duration::from_millis(delay_ms)).await;
```

Comment:
@jkitman @elsirion If my understanding is correct, this is actually pretty big footgun that we've overlooked in the existing code. If this `sleep` is outside of the future we spawn, then we are not doing `futures.next().await` which means other futures are not making progress.

Reply:
I am also thinking about keeping a per-peer counter, so the exponential increase is per-peer.

Reply:
Right, this needs to be a select, possibly even over all the other futures in parallel too?

Reply:
`futures.await` is a select in parallel

I did the per-peer retry counters as well, BTW.